### ISSUE TITLE
Deleting Bay Input that is attached to a Bay Output crashes VCV

### ISSUE THREAD

**Issue by ltackett** [DaveBenham/VenomModules#27]
**Given that:**
* VCV is running a patch with a Bay Input and Bay Output module, and
* The Bay Output source is the Bay Input, and
* The Bay Input module is deleted

**Expected result:**
* The Bay Input module will be removed from the patch

**Actual result:**
* VCV crashes

---

**System information:**
* Apple M2
* MacOS 14.3.1
* VCV Rack2 v2.5.2

**Crash log** (not sure if this is useful):
```
[359.799 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 11. Stack trace:
15: Rack(fatalSignalHandler(int)+40)
14: libsystem_platform.dylib(_sigtramp+56)
13: libRack.dylib(rack::widget::Widget::step()+64)
12: libRack.dylib(rack::widget::Widget::step()+64)
11: libRack.dylib(rack::widget::Widget::step()+64)
10: libRack.dylib(rack::widget::Widget::step()+64)
9: libRack.dylib(rack::widget::Widget::step()+64)
8: libRack.dylib(rack::widget::Widget::step()+64)
7: libRack.dylib(rack::ui::ScrollWidget::step()+24)
6: libRack.dylib(rack::app::RackScrollWidget::step()+560)
5: libRack.dylib(rack::widget::Widget::step()+64)
4: libRack.dylib(rack::app::Scene::step()+432)
3: libRack.dylib(rack::window::Window::step()+1540)
2: libRack.dylib(rack::window::Window::run()+40)
1: Rack(main+4080)
0: dyld(start+2360)
```

**Comment by cosinekitty** [comment]
I am able to reproduce this on Debian Linux. I haven't been able to make it happen on Windows yet. The stack trace is below. @DaveBenham let me know if I can help diagnose or test stuff for you on Linux.

```
[8.628 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 6. Stack trace:
23: /home/don/Rack/2.5.2/Rack() [0x404c5b]
22: /lib/x86_64-linux-gnu/libc.so.6(+0x38dd0)
21: /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x141)
20: /lib/x86_64-linux-gnu/libc.so.6(abort+0x123)
19: ./libRack.so(+0x2a3de2)
18: ./libRack.so(__cxxabiv1::__terminate(void (*)())+0xa)
17: ./libRack.so(+0x7e2175)
16: ./libRack.so(+0x7e22c7)
15: ./libRack.so(+0x2a3d82)
14: ./libRack.so(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x6a)
13: /home/don/.local/share/Rack2/plugins-lin-x64/Venom/plugin.so(BayOutputModuleWidget::step()+0x5c)
12: ./libRack.so(rack::widget::Widget::step()+0x36)
11: ./libRack.so(rack::widget::Widget::step()+0x36)
10: ./libRack.so(rack::widget::Widget::step()+0x36)
9: ./libRack.so(rack::widget::Widget::step()+0x36)
8: ./libRack.so(rack::widget::Widget::step()+0x36)
7: ./libRack.so(rack::ui::ScrollWidget::step()+0x13)
6: ./libRack.so(rack::app::RackScrollWidget::step()+0x36d)
5: ./libRack.so(rack::widget::Widget::step()+0x36)
4: ./libRack.so(rack::window::Window::step()+0x595)
3: ./libRack.so(rack::window::Window::run()+0x28)
2: /home/don/Rack/2.5.2/Rack(main+0xe98)
1: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xea)
0: /home/don/Rack/2.5.2/Rack() [0x404b7a]
```

**Comment by cosinekitty** [comment]
Update: I added some `INFO` statements for debug logging:

```c++
 void widgetStep() { // propagate src labels to output
    for (int i=0; i<OUTPUTS_LEN; i++) {
      PortInfo* oi = outputInfos[i];
      PortExtension* oe = &outputExtensions[i];
      INFO("widgetStep: i=%d/%d, oi=%p", i, OUTPUTS_LEN, oi);
      bool propagate = (oi->name == oe->factoryName);
      INFO("    propagate = %d, srcMod = %p", static_cast<int>(propagate), srcMod);
      if (srcMod)
      {
        INFO("    i=%d, srcMod->inputInfos.size() = %d", i, (int)srcMod->inputInfos.size());
        INFO("    srcMod->inputInfos[%d] = %p", i, srcMod->inputInfos[i]);
        oe->factoryName = srcMod->inputInfos[i]->name;
      }
      else
        oe->factoryName = string::f("Port %d", i+1);
      if (propagate)
        oi->name = oe->factoryName;
      if (bayOutputType) {
        PortInfo* ii = inputInfos[i];
        PortExtension* ie = &inputExtensions[i];
        propagate = (ii->name == ie->factoryName);
        ie->factoryName = oi->name + " normal";
        if (propagate)
          ii->name = ie->factoryName;
      }  
    }
  }
```

To me it looks like `srcMod` is a non-null pointer, but it's pointing at garbage memory. The instant I delete the Bay Input module, this happens:

```
[3.928 info src/BayModule.hpp:195 widgetStep] widgetStep: i=0/8, oi=0x15d0700
[3.928 info src/BayModule.hpp:197 widgetStep]     propagate = 1, srcMod = 0xde8250
[3.928 info src/BayModule.hpp:200 widgetStep]     i=0, srcMod->inputInfos.size() = 8
[3.928 info src/BayModule.hpp:201 widgetStep]     srcMod->inputInfos[0] = 0x7fb9e91d6f20
[3.932 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 6. Stack trace:
23: /home/don/Rack/2.5.2/Rack() [0x404c5b]
22: /lib/x86_64-linux-gnu/libc.so.6(+0x38dd0)
21: /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x141)
20: /lib/x86_64-linux-gnu/libc.so.6(abort+0x123)
19: ./libRack.so(+0x2a3de2)
18: ./libRack.so(__cxxabiv1::__terminate(void (*)())+0xa)
17: ./libRack.so(+0x7e2175)
16: ./libRack.so(+0x7e22c7)
15: ./libRack.so(+0x2a3d82)
14: ./libRack.so(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x6a)
13: /home/don/.local/share/Rack2/plugins-lin-x64/Venom/plugin.so(BayOutputModuleWidget::step()+0x169)
12: ./libRack.so(rack::widget::Widget::step()+0x36)
11: ./libRack.so(rack::widget::Widget::step()+0x36)
10: ./libRack.so(rack::widget::Widget::step()+0x36)
9: ./libRack.so(rack::widget::Widget::step()+0x36)
8: ./libRack.so(rack::widget::Widget::step()+0x36)
7: ./libRack.so(rack::ui::ScrollWidget::step()+0x13)
6: ./libRack.so(rack::app::RackScrollWidget::step()+0x36d)
5: ./libRack.so(rack::widget::Widget::step()+0x36)
4: ./libRack.so(rack::window::Window::step()+0x595)
3: ./libRack.so(rack::window::Window::run()+0x28)
2: /home/don/Rack/2.5.2/Rack(main+0xe98)
1: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xea)
0: /home/don/Rack/2.5.2/Rack() [0x404b7a]
```

Compare ` srcMod->inputInfos[0] = 0x7fb9e91d6f20` here, with how the pointer looks in a normal case, before the Bay Input is deleted.  The valid inputInfos[i] pointers look like 0x15d0400, 0x15d0460, 0x15d05e0, etc. There is a nice pattern until I delete the Bay Input. Then it turns into a totally different pointer 0x7fb9e91d6f20. See this excerpt from earlier in the log, before the Bay Input is deleted:

```
[3.911 info src/BayModule.hpp:195 widgetStep] widgetStep: i=0/8, oi=0x15d0700
[3.911 info src/BayModule.hpp:197 widgetStep]     propagate = 1, srcMod = 0xde8250
[3.911 info src/BayModule.hpp:200 widgetStep]     i=0, srcMod->inputInfos.size() = 8
[3.911 info src/BayModule.hpp:201 widgetStep]     srcMod->inputInfos[0] = 0x15d0400
[3.911 info src/BayModule.hpp:195 widgetStep] widgetStep: i=1/8, oi=0x15d0b30
[3.911 info src/BayModule.hpp:197 widgetStep]     propagate = 1, srcMod = 0xde8250
[3.911 info src/BayModule.hpp:200 widgetStep]     i=1, srcMod->inputInfos.size() = 8
[3.911 info src/BayModule.hpp:201 widgetStep]     srcMod->inputInfos[1] = 0x15d0460
[3.911 info src/BayModule.hpp:195 widgetStep] widgetStep: i=2/8, oi=0x15d0b90
[3.911 info src/BayModule.hpp:197 widgetStep]     propagate = 1, srcMod = 0xde8250
[3.911 info src/BayModule.hpp:200 widgetStep]     i=2, srcMod->inputInfos.size() = 8
[3.911 info src/BayModule.hpp:201 widgetStep]     srcMod->inputInfos[2] = 0x15d04c0
[3.911 info src/BayModule.hpp:195 widgetStep] widgetStep: i=3/8, oi=0x15d0bf0
[3.911 info src/BayModule.hpp:197 widgetStep]     propagate = 1, srcMod = 0xde8250
[3.911 info src/BayModule.hpp:200 widgetStep]     i=3, srcMod->inputInfos.size() = 8
[3.911 info src/BayModule.hpp:201 widgetStep]     srcMod->inputInfos[3] = 0x15d0520
[3.911 info src/BayModule.hpp:195 widgetStep] widgetStep: i=4/8, oi=0x15d0c50
[3.911 info src/BayModule.hpp:197 widgetStep]     propagate = 1, srcMod = 0xde8250
[3.911 info src/BayModule.hpp:200 widgetStep]     i=4, srcMod->inputInfos.size() = 8
[3.911 info src/BayModule.hpp:201 widgetStep]     srcMod->inputInfos[4] = 0x15d0580
[3.911 info src/BayModule.hpp:195 widgetStep] widgetStep: i=5/8, oi=0x15d0cb0
[3.911 info src/BayModule.hpp:197 widgetStep]     propagate = 1, srcMod = 0xde8250
[3.911 info src/BayModule.hpp:200 widgetStep]     i=5, srcMod->inputInfos.size() = 8
[3.911 info src/BayModule.hpp:201 widgetStep]     srcMod->inputInfos[5] = 0x15d05e0
[3.911 info src/BayModule.hpp:195 widgetStep] widgetStep: i=6/8, oi=0x15d0d10
[3.911 info src/BayModule.hpp:197 widgetStep]     propagate = 1, srcMod = 0xde8250
[3.911 info src/BayModule.hpp:200 widgetStep]     i=6, srcMod->inputInfos.size() = 8
[3.911 info src/BayModule.hpp:201 widgetStep]     srcMod->inputInfos[6] = 0x15d0640
[3.911 info src/BayModule.hpp:195 widgetStep] widgetStep: i=7/8, oi=0x15d0d70
[3.911 info src/BayModule.hpp:197 widgetStep]     propagate = 1, srcMod = 0xde8250
[3.911 info src/BayModule.hpp:200 widgetStep]     i=7, srcMod->inputInfos.size() = 8
[3.911 info src/BayModule.hpp:201 widgetStep]     srcMod->inputInfos[7] = 0x15d06a0
```

A theory for  your consideration: you update `srcMod` in `BayOutputModule::process()`. I don't think it ever gets set back to NULL anywhere else. But you are accessing this pointer from `BayOutputModule::widgetStep`, which is running in a different thread: the UI display thread, not the audio thread.

I'm pretty sure VCV Rack does the module deletion in its UI thread. The process method hasn't been called yet to clean up srcMod, so it still points into the memory that just got freed. It accesses invalid memory and crashes trying to do something with a std::string.

I think your process method might be able to clear srcMod back to NULL, but it hasn't had a chance to run yet, so srcMod is still left pointing at a module that has been destroyed.

**Comment by DaveBenham** [comment]
Thanks for the report @ltackett 

And wow - thanks for the bug tracing @cosinekitty !

That makes sense. So I think I can fix that by simply moving the label propagation to the BayOutputModule::process(). I just need to not do that work every frame else it will impact performance.

**Comment by DaveBenham** [comment]
@cosinekitty - I think I may have fixed the problem. Can you test the latest build in my Actions - build #499.

The fixed code is in my dev branch.

**Comment by ltackett** [comment]
@cosinekitty ftw!

@DaveBenham I’ll also build latest and test.

**Comment by ltackett** [comment]
Fix is confirmed working for me. I’ll wait to hear from @cosinekitty before closing. Thanks so much for the quick work @DaveBenham  and @cosinekitty !

**Comment by cosinekitty** [comment]
Yes, that fixes the crash for me too!

However, it seems a little risky to do memory allocations (the std::string stuff) in the process method. Heap allocations are unbounded in execution time which could cause audio stuttering. Would it be just as easy to do all this stuff from the UI thread (in a `step` method)?

On the other hand, I'm assuming string copy/format stuff only happens only when someone manually changes a label?

**Comment by DaveBenham** [comment]
Good points.

With the fix, the labels are propagated once every 32 sample cycles. Unfortunately, the code blindly copies the strings each time. I can/should/will add a condition to only copy if a change is detected.

As for changing the location of this processing, that would be tricky at this point. It all ties into how I manage port/control names and locking for the entire plugin. I might revisit this in the future.

This fix will appear in the library as part of v2.10, which I am planning for later this month.

I'm closing this issue. Thanks again!
